### PR TITLE
fix(lv_pos): fix invalidate area problem with transform

### PR DESCRIPTION
**Pull request by W-Mai**
### Description of the feature or fix

Bug cause: Failure to store the area of the previous transform operation, failing to invalidate the region of the previous transform when a new transform operation occurs.

Also, unnecessary +5 operations were removed from `lv_obj_get_transformed_area` function.

Before:
image

After:
image

### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `LV_ITERATE_ROOTS` list in [`lv_gc.h`](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_gc.h) and mark the variable with `GC_ROOT(variable)` when it's used. See a detaild description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<modul_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the followings needs to be followed (see a detaild description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)): 
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Comment by kisvegabor** [comment]
What if we simply add `lv_obj_invalidate` at the beginning of `lv_obj_add_style()` and `lv_obj_remove_style()`.

In `lv_obj_set_local_style_prop` we can also invalidate before setting the property if the property is transformation related. 

The only thing that we can't manage is when a style is already assigned to some objects, the style properties are changed and `lv_obj_report_style_change` is called. In this case the style is already changed and we can't get the previous transformed coordinates. What we can do is calling `lv_obj_report_style_change` before changing the style too.

What do you think?

**Comment by W-Mai** [comment]
> In `lv_obj_set_local_style_prop`, we can also invalidate before setting the property if the property is transformation related.

If we manually invalidate, as there is no cached result of the transformed style, we need to recalculate the coordinator, which will waste a lot of time.

**Comment by kisvegabor** [comment]
Is it really a lot of time? Compared to drawing a transformed widget, I think it's negligible.

**Comment by W-Mai** [comment]
> Is it really a lot of time? Compared to drawing a transformed widget, I think it's negligible.

like this?

```cpp
void lv_obj_set_local_style_prop(lv_obj_t * obj, lv_style_prop_t prop, lv_style_value_t value,
                                 lv_style_selector_t selector)
{
    lv_style_t * style = get_local_style(obj, selector);
    lv_obj_invalidate(obj);
    lv_style_set_prop(style, prop, value);
    lv_obj_refresh_style(obj, selector, prop);
}
```

**Comment by FASTSHIFT** [comment]
> Is it really a lot of time? Compared to drawing a transformed widget, I think it's negligible.

`lv_obj_get_transformed_area` takes around 30~40us to execute on a 288MHz processor.

**Comment by W-Mai** [comment]
> like this?
> 
> ```c++
> void lv_obj_set_local_style_prop(lv_obj_t * obj, lv_style_prop_t prop, lv_style_value_t value,
>                                  lv_style_selector_t selector)
> {
>     lv_style_t * style = get_local_style(obj, selector);
>     lv_obj_invalidate(obj);
>     lv_style_set_prop(style, prop, value);
>     lv_obj_refresh_style(obj, selector, prop);
> }
> ```

If you follow this approach, it may double the transform time since `lv_obj_refresh_style` will trigger 2 `lv_obj_invalidate` calls, and each `lv_obj_invalidate` call can potentially cause at least 2 `lv_obj_get_transformed_area` calls.

As a result, the total number of `lv_obj_get_transformed_area` calls can be up to 4. However, if you record the transformed area, you can use `_lv_inv_area` to save time and avoid multiple calls to `lv_obj_get_transformed_area`.

I think that time is more valuable. What do you think?

**Comment by kisvegabor** [comment]
So it's 35 us x 288 MHz = ~10k instructions. I don't really see why it needs so many instructions. Can you use a profiler and call this function 1000 times to see why does it take so long?

**Comment by FASTSHIFT** [comment]
> So it's 35 us x 288 MHz = ~10k instructions. I don't really see why it needs so many instructions. Can you use a profiler and call this function 1000 times to see why does it take so long?

I tested it again, for this scenario:
```c
lv_obj_t* obj = lv_obj_create(lv_scr_act());
lv_obj_set_style_transform_angle(obj, 450, 0);
```

```bash
[User] lv_obj_get_transformed_area: cost 1285us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1285us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1285us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1285us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1285us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1285us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1313us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1313us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1313us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1285us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 23986us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 1313us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 23309us 	(in lv_obj_pos.c line #820)
[User] lv_obj_get_transformed_area: cost 23402us 	(in lv_obj_pos.c line #820)
```
Call `lv_obj_get_transformed_area` (average of 1000 times) each time most of the time is about 1.3us, the longest time is about 23us.

**Comment by FASTSHIFT** [comment]
The following is a profiler analysis:
image

**Comment by kisvegabor** [comment]
Very useful, thank you very much!

> Call lv_obj_get_transformed_area (average of 1000 times) each time most of the time is about 1.3us, the longest time is about 23us.

1.3us sounds much better than 35us. :slightly_smiling_face: Do you know what are those 23 us peeks? Can it be related to LVGL?

From the profiler result I concluded that `lv_obj_get_style_prop` needs to be optimized as it has the highest self time (I already have some ideas to optimize it).

What is your conclusion?

**Comment by FASTSHIFT** [comment]
> Very useful, thank you very much!
> 
> > Call lv_obj_get_transformed_area (average of 1000 times) each time most of the time is about 1.3us, the longest time is about 23us.
> 
> 1.3us sounds much better than 35us. 🙂 Do you know what are those 23 us peeks? Can it be related to LVGL?
> 
> From the profiler result I concluded that `lv_obj_get_style_prop` needs to be optimized as it has the highest self time (I already have some ideas to optimize it).
> 
> What is your conclusion?

I re-measured the CPU cycle data consumed by `lv_obj_get_style_prop` and found that it takes about 0.8\~3us. In one `lv_obj_get_transformed_area` call (if obj is set to transform), `lv_obj_get_style_prop` needs to be called 16 times. So it can be considered that `lv_obj_get_style_prop` takes up most of the time.

We need to optimize it, because this function is used very frequently, `lv_obj_get_style_prop` needs to be called about 1000\~2000 times in one rendering of `lv_demo_widgets`, which means that the total time for LVGL to read the style attribute is about 2\~3ms, which cannot be ignored performance overhead.

The following is some test data intercepted:

`lv_obj_get_style_prop` takes time:
```bash
lv_obj_get_style_prop: cycle = 850 (2.951us)
lv_obj_get_style_prop: cycle = 229 (0.795us)
lv_obj_get_style_prop: cycle = 232 (0.806us)
lv_obj_get_style_prop: cycle = 229 (0.795us)
lv_obj_get_style_prop: cycle = 229 (0.795us)
lv_obj_get_style_prop: cycle = 232 (0.806us)
lv_obj_get_style_prop: cycle = 655 (2.274us)
lv_obj_get_style_prop: cycle = 655 (2.274us)
lv_obj_get_style_prop: cycle = 229 (0.795us)
lv_obj_get_style_prop: cycle = 229 (0.795us)
lv_obj_get_style_prop: cycle = 232 (0.806us)
lv_obj_get_style_prop: cycle = 229 (0.795us)
lv_obj_get_style_prop: cycle = 229 (0.795us)
lv_obj_get_style_prop: cycle = 232 (0.806us)
lv_obj_get_style_prop: cycle = 229 (0.795us)
lv_obj_get_style_prop: cycle = 229 (0.795us)
lv_obj_get_style_prop: cycle = 412 (1.431us)
lv_obj_get_style_prop: cycle = 614 (2.132us)
lv_obj_get_style_prop: cycle = 543 (1.885us)
lv_obj_get_style_prop: cycle = 495 (1.719us)
lv_obj_get_style_prop: cycle = 958 (3.326us)
lv_obj_get_style_prop: cycle = 1023 (3.552us)
lv_obj_get_style_prop: cycle = 1023 (3.552us)
lv_obj_get_style_prop: cycle = 1023 (3.552us)
lv_obj_get_style_prop: cycle = 412 (1.431us)
lv_obj_get_style_prop: cycle = 412 (1.431us)
lv_obj_get_style_prop: cycle = 1023 (3.552us)
lv_obj_get_style_prop: cycle = 559 (1.941us)
lv_obj_get_style_prop: cycle = 583 (2.024us)
lv_obj_get_style_prop: cycle = 1046 (3.632us)
lv_obj_get_style_prop: cycle = 607 (2.108us)
lv_obj_get_style_prop: cycle = 839 (2.913us)
lv_obj_get_style_prop: cycle = 535 (1.858us)
lv_obj_get_style_prop: cycle = 751 (2.608us)
lv_obj_get_style_prop: cycle = 548 (1.903us)
lv_obj_get_style_prop: cycle = 548 (1.903us)
lv_obj_get_style_prop: cycle = 836 (2.903us)
```

The number of calls to `lv_obj_get_style_prop` during the execution of `_lv_disp_refr_timer`:
```bash
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1625 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1938 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1863 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1764 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2293 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1581 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1463 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1769 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1863 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2004 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1797 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2048 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1839 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1797 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1938 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1764 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1771 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1463 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1844 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1764 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1962 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1891 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1839 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2037 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2146 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1865 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1834 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1875 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1505 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1938 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1971 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1896 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2161 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1463 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1797 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2070 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1797 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2147 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2023 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1839 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 2004 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1764 	(in lv_refr.c line #348)
[User] _lv_disp_refr_timer: get_style_prop_cnt = 1929 	(in lv_refr.c line #348)
```

**Comment by kisvegabor** [comment]
We should cut the execution time of `lv_obj_get_style_prop` at least to half. An we should do in v9, before now we can still do "anything" to speed it up.

---

Going back to the original topic of this PR, it seems [my suggestion](https://github.com/lvgl/lvgl/pull/4228#issuecomment-1552872647) won't cause a bottleneck and we should optimize `lv_obj_get_style_prop` independently.

I light of that what do you think about my proposed solution?

**Comment by W-Mai** [comment]
In fact, using lv_obj_invalidate directly is very concise. However, there may be precision issues with lv_obj_get_transformed_area itself. In my testing, there may be a difference of 1 pixel at 90 and 180 degrees. I don't want to expand the area by adding +1 or other methods in lv_obj_get_transformed_area. So, this simple modification method may have a remaining one-pixel issue. Is there any better solution?

**Comment by kisvegabor** [comment]
I suspect that the rounding error is because of the last element of `sin0_90_table` in `lv_math.c` is `32767` but it should be `32768` to get `1` after `>> LV_TRIGO_SHIFT`. 

Could you try it out? Note that after this `lv_trigo_sin` should return `int32_t` and `sin0_90_table` should be `int32_t`.

**Comment by W-Mai** [comment]
> I suspect that the rounding error is because of the last element of `sin0_90_table` in `lv_math.c` is `32767` but it should be `32768` to get `1` after `>> LV_TRIGO_SHIFT`.
> 
> Could you try it out? Note that after this `lv_trigo_sin` should return `int32_t` and `sin0_90_table` should be `int32_t`.

I'll try it and give you the result later 🤓

**Comment by W-Mai** [comment]
image

My problem still exists

I will try to find out the issue. If I can't find it, I will write a simple demo to reproduce the issue and we can discuss it together.

```cpp
LV_ATTRIBUTE_FAST_MEM int32_t lv_trigo_sin(int16_t angle)
{
    int32_t ret = 0;
    angle       = angle % 360;
```

```cpp
static const int32_t sin0_90_table[] = {
    0,     572,   1144,  1715,  2286,  2856,  3425,  3993,  4560,  5126,  5690,  6252,  6813,  7371,  7927,  8481,
    9032,  9580,  10126, 10668, 11207, 11743, 12275, 12803, 13328, 13848, 14364, 14876, 15383, 15886, 16383, 16876,
    17364, 17846, 18323, 18794, 19260, 19720, 20173, 20621, 21062, 21497, 21925, 22347, 22762, 23170, 23571, 23964,
    24351, 24730, 25101, 25465, 25821, 26169, 26509, 26841, 27165, 27481, 27788, 28087, 28377, 28659, 28932, 29196,
    29451, 29697, 29934, 30162, 30381, 30591, 30791, 30982, 31163, 31335, 31498, 31650, 31794, 31927, 32051, 32165,
    32269, 32364, 32448, 32523, 32587, 32642, 32687, 32722, 32747, 32762, 32768
};
```

**Comment by W-Mai** [comment]
> I will try to find out the issue. If I can't find it, I will write a simple demo to reproduce the issue and we can discuss it together.

Okay, I finally found the cause of the problem. 
In `_lv_img_buf_get_transformed_area` function, it forgot to decrease the passed width and height by 1

However, I noticed that when trying to solve this problem earlier, I removed the "-1" operation as shown in [https://forum.lvgl.io/t/inaccurate-bitmap-zoom/8858](https://forum.lvgl.io/t/inaccurate-bitmap-zoom/8858), which eventually caused the current problem. One possible reason for this problem is that I previously removed the "+2" and "-2" operations in "_lv_img_buf_get_transformed_area", but did not handle the "width-1" and "height-1" above it. 🤔️

At the same time, it is not enough to just modify the last digit of the sin table. I remapped the sin table from 0 to 32768 using the following code:

```python
my_table = []
for i in range(0, 91):
    angle = i / 180 * pi
    s = sin(angle) * 32768

    my_table.append(round(s))

print("{", ", ".join(map(str, my_table)), "}")
```

**Comment by kisvegabor** [comment]
Finally with all these modifications does it work correctly?

**Comment by W-Mai** [comment]
> Finally with all these modifications does it work correctly?

In my case it works correctly.
Maybe need more tests?

**Comment by kisvegabor** [comment]
Please commit your changes so that I can also test them.

**Comment by W-Mai** [comment]
> Please commit your changes so that I can also test them.

I've submitted these commits 5h ago😍

**Comment by kisvegabor** [comment]
Oooo, sorry :) I'll check it later today.

File: src/core/lv_obj_style.c [hunk 1]
```diff
@@ -330,6 +330,7 @@ void lv_obj_set_local_style_prop(lv_obj_t * obj, lv_style_prop_t prop, lv_style_
                                  lv_style_selector_t selector)
 {
     lv_style_t * style = get_local_style(obj, selector);
+    lv_obj_invalidate(obj);
```
Comment (kisvegabor) [inline src/core/lv_obj_style.c]:
```suggestion
    if(part == LV_PART_MAIN && lv_style_prop_has_flag(prop, LV_STYLE_PROP_TRANSFORM); lv_obj_invalidate(obj);
```

`LV_STYLE_PROP_TRANSFORM` needs to be added in `lv_style.c/h`.

**Review by kisvegabor — COMMENTED** [review]
It looks great! It's so great that it's finally fixed.


Just one optimization idea above.

We should also add an invalidation at the beginning of `lv_obj_add/remove_style`.

**Comment by W-Mai** [comment]
done, I've added these code in lv_obj_add/remove/set_local_style

File: src/core/lv_obj_style.c [hunk 2]
```diff
@@ -924,4 +932,12 @@ static void fade_in_anim_ready(lv_anim_t * a)
     lv_obj_remove_local_style_prop(a->var, LV_STYLE_OPA, 0);
 }
 
-
+static bool is_transform_style(lv_style_prop_t prop)
```
Comment (kisvegabor) [inline src/core/lv_obj_style.c]: Could you handle it by adding a new `LV_STYLE_PROP_FLAG_TRANSFORM` similarly to [these](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_style.c#L41).

**Review by kisvegabor — COMMENTED** [review]
Looks great, just one minor thing :slightly_smiling_face:

File: src/core/lv_obj_style.c (hunk 2 above)
Comment (W-Mai) [reply src/core/lv_obj_style.c]: great, let me do it

Comment (W-Mai) [reply src/core/lv_obj_style.c]: done

**Comment by kisvegabor** [comment]
It looks good thank you!  Do have any idea about the failing tests. They seem unrelated.

Maybe rebasing the PR to master will solve it.

**Comment by W-Mai** [comment]
I don't know why. No matter how I rebase, I can't clean up other people's submissions. I don't know where the problem is. It's so strange.

**Comment by W-Mai** [comment]
I create a new pr https://github.com/lvgl/lvgl/pull/4285

Let's move to that pr

**Pull request closed without merging**